src/gui/dbsettings/DatabaseSettingsWidgetStatistics.cpp

```diff
@@ -0,0 +1,206 @@
+/*
+ *  Copyright (C) 2019 KeePassXC Team <team@keepassxc.org>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 or (at your option)
+ *  version 3 of the License.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "DatabaseSettingsWidgetStatistics.h"
+#include "ui_DatabaseSettingsWidgetStatistics.h"
+
+#include "core/Database.h"
+#include "core/FilePath.h"
+#include "core/Group.h"
+#include "core/Metadata.h"
+
+#include <QFileInfo>
+#include <QMap>
+#include <QMessageBox>
+#include <QStandardItemModel>
+
+namespace
+{
+    class Stats
+    {
+    public:
+        // The statistics we collect:
+        QDateTime modified; // File modification time
+        int nGroups = 0; // Number of groups in the database
+        int nEntries = 0; // Number of entries (across all groups)
+        int nExpired = 0; // Number of expired entries
+        int nPwdsUnique = 0; // Number of unique passwords
+        int nPwdsReused = 0; // Number of non-unique passwords
+        int pwdTotalLen = 0; // Total length of all passwords
+
+        // Ctor does all the work
+        Stats(QSharedPointer<Database> db)
+            : modified(QFileInfo(db->filePath()).lastModified())
+        {
+            countGroup(*db->rootGroup());
+        }
+
+        // Get average password length
+        int averagePwdLength() const
+        {
+            return pwds_.empty() ? 0 : pwdTotalLen / pwds_.size();
+        }
+
+        // Get max number of password reuse (=how many entries
+        // share the same password)
+        int maxPwdReuse() const
+        {
+            int ret = 0;
+            for (const auto& count : pwds_) {
+                ret = std::max(ret, count);
+            }
+            return ret;
+        }
+
+        // A warning sign is displayed if one of the
+        // following returns true.
+        bool isAnyExpired() const
+        {
+            return nExpired > 0;
+        }
+
+        bool areTooManyPwdsReused() const
+        {
+            return nPwdsReused > nPwdsUnique / 10;
+        }
+
+        bool arePwdsReusedTooOften() const
+        {
+            return maxPwdReuse() > 3;
+        }
+
+        bool isAvgPwdTooShort() const
+        {
+            return averagePwdLength() < 10;
+        }
+
+    private:
+        QMap<QString, int> pwds_;
```

Comment:
Minor code style point, we use "m_" prefix for member variables. `m_passwords`
You also want to use a QHash since it is faster.